shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -59,17 +61,156 @@ final class TsTypegenCodeBuilder {
       val typegenTermScope = Scope("localTermScope", List.empty, termScope)
-      TypegenContext(typeVarMapping, typegenTermScope, typegenTypeScope)
+      TypegenContext(typeVarMapping, typegenTermScope, typegenTypeScope, false)
+    }
+  }
+
+  def declareTypeDef(typeDef: TypeDef): TypeSymbol = typeDef match {
+    case TypeDef(Als, TypeName(name), tparams, body, _, _) =>
+      typeScope.declareTypeAlias(name, tparams map { _.name }, body)
+    case TypeDef(Trt, TypeName(name), tparams, body, _, _) =>
+      typeScope.declareTrait(name, tparams map { _.name }, body)
+    case TypeDef(Cls, TypeName(name), tparams, baseType, _, members) =>
+      typeScope.declareClass(name, tparams map { _.name }, baseType, members)
+  }
+
+  /**
+    * Start adding type definition and store partially translated
+    * type definition as state
+    *
+    * @param typeDef
+    */
+  def addTypeDefStart(typeDef: TypeDef): Unit = {
+    if (currentTypedefCode.isDefined) {
+      throw CodeGenError(s"Cannot start adding new type definition for ${typeDef.nme.name} without complete previous")
+    }
+    val tySymb = typeScope.getTypeSymbol(typeDef.nme.name).getOrElse(
+      throw CodeGenError(s"No type definition for ${typeDef.nme.name} exists")
+    )
+    tySymb match {
+      case (classInfo: ClassSymbol) => addTypeGenClassDef(classInfo)
+      case (aliasInfo: TypeAliasSymbol) => addTypeGenTypeAlias(aliasInfo)
+      case (traitInfo: TraitSymbol) => throw CodeGenError("Typegen for traits is not supported currently")
+    }
+  }
+
+  /** Add class method type gen translation to the partially translated
+    * class typegen code
+    */
+  def addClassMethods(methodName: String, methodType: Type): Unit = {
+    val (className, currentDef) = currentTypedefCode.getOrElse(throw CodeGenError(s"Cannot add method without starting a type definition"))
+    val classInfo = typeScope.getClassSymbol(className).getOrElse(throw CodeGenError(s"Cannot find class $className for which method $methodName is being declared"))
+    val classTypeParams = classInfo.params.toSet
+    // unwrap method function type to remove implicit this
+    val methodBodyType = methodType match {
+      case Function(lhs, rhs) => rhs
+      case _ => throw CodeGenError(s"Cannot translate malformed method: $methodName because it does not have top level function")
+    }
+    // Create a mapping from type var to their friendly name for lookup
+    val typegenCtx = TypegenContext(methodBodyType, true)
+    val tsType = toTsType(methodBodyType)(typegenCtx, Some(true));
+    // only use free variables for type parameters
+    val typeParams = typegenCtx.typeVarMapping.iterator
+      .filter(tup => methodBodyType.freeTypeVariables.contains(tup._1) &&
+        // ignore class type parameters since they are implicitly part of class scope
+        tup._1.nameHint.fold(false)(!classTypeParams.contains(_))
+      )
+      .map(_._2)
+      .toList
```

Comment:
```suggestion
      .toList.sorted
```

Also do this in `def` type parameter lists.

Reply:
Storing in SortedMap by defining an ordering on TypeVars.

This is not needed for class type parameters, since they are never collected into a map so their original ordering is maintained.

Reply:
But it's not currently stored in a SortedMap, is it?

Reply:
pushed the change